client.go

```diff
@@ -0,0 +1,354 @@
+				ResourceNamesSubscribe: slices.Collect(subs.Values()),
+			})
+			if err != nil {
+				return err
+			}
+		}
+		return nil
+	}
+
+	isFirst := true
+	for {
+		err := sendPendingSubscriptions()
+		if err != nil {
+			return err
+		}
+
+		select {
+		case <-c.newSubscription:
+			err := sendPendingSubscriptions()
+			if err != nil {
+				return err
+			}
+		case responses := <-responsesCh:
+			h, ok := c.getResourceHandler(responses[0].TypeUrl)
+			if !ok {
+				for _, res := range responses {
+					err := c.sendACKOrNACK(
+						stream,
+						res,
+						fmt.Errorf("received response with unknown type: %q", res.TypeUrl),
+					)
+					if err != nil {
+						slog.WarnContext(stream.Context(), "ADS stream closed", "err", err)
+						return err
+					}
+				}
+				continue
+			}
+
+			// Always ACK all but the last response. Errors will only be reported back to the server once all
+			// chunks are processed.
+			for _, res := range responses[:len(responses)-1] {
+				err := c.sendACKOrNACK(stream, res, nil)
+				if err != nil {
+					return err
+				}
+			}
+
+			handlerErr := h.HandleResponses(isFirst, responses)
+			isFirst = false
+			if err = c.sendACKOrNACK(stream, responses[len(responses)-1], handlerErr); err != nil {
+				return err
+			}
+		case <-stream.Context().Done():
+			return stream.Context().Err()
+		}
+	}
+}
+
+// sendACKOrNACK will send an ACK or NACK (depending on the given error) for the given response.
+func (c *ADSClient) sendACKOrNACK(stream deltaClient, res *ads.DeltaDiscoveryResponse, err error) error {
+	req := &ads.DeltaDiscoveryRequest{
+		Node:          c.node,
+		TypeUrl:       res.TypeUrl,
+		ResponseNonce: res.Nonce,
+	}
+	if err != nil {
+		req.ErrorDetail = status.New(codes.InvalidArgument, err.Error()).Proto()
+		slog.WarnContext(stream.Context(), "NACKing response", "res", res, "err", err)
+	} else {
+		slog.DebugContext(stream.Context(), "ACKing response", "res", res)
+	}
+	return stream.Send(req)
+}
+
+// newStream acquires a fresh stream from getDeltaClient and kicks off a goroutine that will read all
+// responses from the stream, writing them to the returned channel. The goroutine will exit when the
+// stream ends.
+func (c *ADSClient) newStream() (deltaClient, <-chan []*ads.DeltaDiscoveryResponse, error) {
+	stream, err := c.getDeltaClient()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	responses := make(chan []*ads.DeltaDiscoveryResponse)
+	go func() {
+		chunkedResponses := make(map[string][]*ads.DeltaDiscoveryResponse)
+
+		for {
+			res, err := stream.Recv()
+			if err != nil {
+				slog.WarnContext(stream.Context(), "ADS stream closed", "err", err)
+				return
+			}
+
+			slog.Debug("Response received", "res", res)
+
+			var resSlice []*ads.DeltaDiscoveryResponse
+
+			if c.responseChunkingSupported {
+				resSlice = chunkedResponses[res.TypeUrl]
+				resSlice = append(resSlice, res)
+				chunkedResponses[res.TypeUrl] = resSlice
+				if c.responseChunkingSupported {
```

Comment:
still inside the if-check on line 303, no need to re-check

Reply:
Ah yes, I think I must have messed up during refactoring this